File: src/parser/ast.rs
use serde::Deserialize;

#[derive(Debug, Deserialize, Default)]
pub struct Meta {
    pub theme: Option<String>,
    pub title: Option<String>,
    pub footer: Option<String>,
    pub paginate: Option<bool>,
    pub size: Option<String>,
    pub style: Option<String>,
    pub logo: Option<String>,
}

impl Meta {
    pub fn dimensions(&self) -> (u32, u32) {
        match self.size.as_deref().unwrap_or("16:9") {
            "4:3" => (1024, 768),
            "16:10" => (1280, 800),
            s if s.contains('x') => {
                if let Some((w, h)) = s.split_once('x') {
                    if let (Ok(w), Ok(h)) = (w.parse::<u32>(), h.parse::<u32>()) {
                        return (w.min(7680).max(320), h.min(7680).max(320));
                    }
                }
                (1280, 720)
            }
            _ => (1280, 720),
        }
    }
}

#[derive(Debug)]
pub struct Document {
    pub meta: Meta,
    pub slides: Vec<Slide>,
}

#[derive(Debug)]
pub struct Slide {
    pub layout: LayoutType,
    pub children: Vec<Node>,
    pub notes: String,
}

#[derive(Debug, PartialEq)]
pub enum LayoutType { Title, Grid2, Grid3, Grid4, Content }

#[derive(Debug)]
pub enum Node {
    Heading(Heading),
    Paragraph(Vec<Inline>),
    Grid(Grid),
    Card(Card),
    Table(Table),
    Quote(Vec<Inline>),
    List(Vec<String>),
    Attr(AttrNode),
}

#[derive(Debug)]
pub struct Heading { pub level: u8, pub content: Vec<Inline> }

#[derive(Debug)]
pub struct Grid { pub cols: usize, pub class: String, pub children: Vec<Node> }

#[derive(Debug)]
pub struct Card { pub header: String, pub body: Vec<String>, pub tag: Option<String>, pub class: String }

#[derive(Debug)]
pub struct Table { pub headers: Vec<String>, pub rows: Vec<Vec<String>> }

#[derive(Debug)]
pub struct AttrNode { pub typ: String, pub value: String, pub attrs: Vec<(String, String)> }

#[derive(Debug)]
pub enum Inline { Text(String), Strong(Vec<Inline>), Code(String), Link(String, String), SoftBreak }

File: src/slidr/parser/ast.py
"""AST node types for slidr."""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import Optional


class LayoutType(Enum):
    TITLE = "title"
    GRID2 = "grid-2"
    GRID3 = "grid-3"
    GRID4 = "grid-4"
    CONTENT = "content"


@dataclass
class Meta:
    theme: str = ""
    title: str = ""
    footer: str = ""
    paginate: bool = False
    size: str = "16:9"
    style: str = ""
    logo: str = ""
    pygments_style: str = "default"
    seaborn_theme: str = "muted"
    theme_variant: str = "light"

    def dimensions(self) -> tuple[int, int]:
        raw = str(self.size)
        if "x" in raw:
            w, h = raw.split("x")
            return (min(max(int(w), 320), 7680), min(max(int(h), 320), 7680))
        return {"16:9": (1280, 720), "4:3": (1024, 768), "16:10": (1280, 800)}.get(raw, (1280, 720))

    def physical_dims(self) -> tuple[float, float]:
        """Slide dimensions in mm for ODP/print."""
        raw = str(self.size)
        return {"16:9": (280.0, 157.5), "4:3": (280.0, 210.0),
                "16:10": (280.0, 175.0)}.get(raw, (280.0, 157.5))


@dataclass
class Document:
    meta: Meta
    slides: list["Slide"]


@dataclass
class Slide:
    layout: str = "content"
    children: list["Node"] = field(default_factory=list)
    notes: str = ""
    variant: str = ""


class Node:
    pass


@dataclass
class Heading(Node):
    level: int = 1
    content: list["Inline"] = field(default_factory=list)


@dataclass
class Paragraph(Node):
    content: list["Inline"] = field(default_factory=list)


@dataclass
class Grid(Node):
    cols: int = 2
    class_: str = ""
    children: list[Node] = field(default_factory=list)


@dataclass
class Card(Node):
    header: str = ""
    body: list[str] = field(default_factory=list)
    tag: Optional[str] = None
    class_: str = ""


@dataclass
class Arrow(Node):
    """Connector between two cards (arrow icon, label, etc.)."""
    content: str = ""


@dataclass
class Notes(Node):
    """Full-width conclusion card below a comparison layout."""
    content: str = ""
    tag: Optional[str] = None


@dataclass
class Row(Node):
    """Horizontal row of elements placed side by side."""
    children: list[Node] = field(default_factory=list)


@dataclass
class Table(Node):
    headers: list[str] = field(default_factory=list)
    rows: list[list[str]] = field(default_factory=list)


@dataclass
class Quote(Node):
    content: list["Inline"] = field(default_factory=list)


@dataclass
class ListNode(Node):
    items: list[list[Inline]] = field(default_factory=list)


@dataclass
class AttrNode(Node):
    type: str = ""
    value: str = ""
    attrs: dict[str, str] = field(default_factory=dict)


@dataclass
class CodeBlock(Node):
    content: str = ""
    language: str = ""


@dataclass
class Inline:
    pass


@dataclass
class Text(Inline):
    content: str = ""


@dataclass
class Strong(Inline):
    children: list[Inline] = field(default_factory=list)


@dataclass
class Emphasis(Inline):
    children: list[Inline] = field(default_factory=list)


@dataclass
class Strikethrough(Inline):
    children: list[Inline] = field(default_factory=list)


@dataclass
class CodeSpan(Inline):
    content: str = ""


@dataclass
class Image(Inline):
    src: str = ""
    alt: str = ""
    title: str = ""


@dataclass
class SoftBreak(Inline):
    pass
